apps/opencs/view/render/terraintexturemode.cpp

```diff
@@ -0,0 +1,328 @@
+    connect(textureBrushWindow, SIGNAL(passBrushShape(int)), this, SLOT(setBrushShape(int)));
+}
+
+void CSVRender::TerrainTextureMode::activate(CSVWidget::SceneToolbar* toolbar)
+{
+    EditMode::activate(toolbar);
+}
+
+void CSVRender::TerrainTextureMode::deactivate(CSVWidget::SceneToolbar* toolbar)
+{
+    EditMode::deactivate(toolbar);
+}
+
+void CSVRender::TerrainTextureMode::primaryEditPressed(const WorldspaceHitResult& hit) // Apply changes here
+{
+    editTerrainTextureGrid(hit);
+}
+
+void CSVRender::TerrainTextureMode::primarySelectPressed(const WorldspaceHitResult& hit)
+{
+      QPoint position = QCursor::pos();
+      textureBrushWindow->move (position);
+      textureBrushWindow->show();
+}
+
+void CSVRender::TerrainTextureMode::secondarySelectPressed(const WorldspaceHitResult& hit)
+{
+}
+
+bool CSVRender::TerrainTextureMode::primaryEditStartDrag (const QPoint& pos)
+{
+    WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+    editTerrainTextureGrid(hit);
+    return false;
+}
+
+bool CSVRender::TerrainTextureMode::secondaryEditStartDrag (const QPoint& pos)
+{
+    return false;
+}
+
+bool CSVRender::TerrainTextureMode::primarySelectStartDrag (const QPoint& pos)
+{
+    return false;
+}
+
+bool CSVRender::TerrainTextureMode::secondarySelectStartDrag (const QPoint& pos)
+{
+    return false;
+}
+
+void CSVRender::TerrainTextureMode::drag (const QPoint& pos, int diffX, int diffY, double speedFactor) {
+}
+
+void CSVRender::TerrainTextureMode::dragCompleted(const QPoint& pos) {
+}
+
+void CSVRender::TerrainTextureMode::dragAborted() {
+}
+
+void CSVRender::TerrainTextureMode::dragWheel (int diff, double speedFactor) {}
+
+void CSVRender::TerrainTextureMode::handleDragEnterEvent (QDragEnterEvent *event) {
+    event->accept();
+}
+
+void CSVRender::TerrainTextureMode::handleDropEvent (QDropEvent *event) {
+  const CSMWorld::TableMimeData* mime = dynamic_cast<const CSMWorld::TableMimeData*> (event->mimeData());
+
+  if (!mime) // May happen when non-records (e.g. plain text) are dragged and dropped
+      return;
+
+  if (mime->holdsType (CSMWorld::UniversalId::Type_LandTexture))
+  {
+      const std::vector<CSMWorld::UniversalId> ids = mime->getData();
+
+      for (const CSMWorld::UniversalId& uid : ids)
+      {
+          mBrushTexture = uid.getId();
+          emit passBrushTexture(mBrushTexture);
+      }
+  }
+}
+
+void CSVRender::TerrainTextureMode::editTerrainTextureGrid(const WorldspaceHitResult& hit)
+{
+    mCellId = getWorldspaceWidget().getCellId (hit.worldPos);
+    std::string hash = "#";
+    std::string space = " ";
+    std::size_t hashlocation = mCellId.find(hash);
+    std::size_t spacelocation = mCellId.find(space);
+    std::string slicedX = mCellId.substr (hashlocation+1, spacelocation-hashlocation);
+    std::string slicedY = mCellId.substr (spacelocation+1);
+    CSMWorld::CellCoordinates mCoordinates(stoi(slicedX), stoi(slicedY));
```

Comment:
This is a very roundabout way of parsing a cell ID. I would just use an std::istringstream. There are probably several examples in the code base.

Reply:
CSMWorld::CellCoordinates CSMWorld::RegionMap::getIndex for example.